### ISSUE TITLE
edition lint: prefer `'_` to lifetime elision in types

### ISSUE THREAD

**Issue by nrc** [rust-lang/rust#52041]
E.g., use `Ref<'_, T>` rather than `Ref<T>`

**Comment by zackmdavis** [comment]
This looks like a duplicate of #45992, implemented in #46254? Some more work is needed, however. As far as polish goes:

 * `elided-lifetimes-in-paths` is not currently listed as an Edition lint
 * it should have a structured suggestion (for RLS, rustfix, _&c._)

More pressingly, it seems to have a lot of false positives. This code:

```rust
#![feature(rust_2018_preview)]
#![warn(elided_lifetimes_in_paths)]

use std::cell::{RefCell, Ref};

fn main() {
    // example from `Ref` documentation:
    // https://doc.rust-lang.org/std/cell/struct.Ref.html#examples
    let c = RefCell::new((5, 'b'));
    let b1: Ref<(u32, char)> = c.borrow();
    let b2: Ref<u32> = Ref::map(b1, |t| &t.0);
    assert_eq!(*b2, 5);
}
```

results in this output:

```
zmd@ReflectiveCoherence:~/Code/Misc$ rustc +nightly elide_ref.rs 
warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref.rs:10:13
   |
10 |     let b1: Ref<(u32, char)> = c.borrow();
   |             ^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> elide_ref.rs:2:9
   |
2  | #![warn(elided_lifetimes_in_paths)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref.rs:11:24
   |
11 |     let b2: Ref<u32> = Ref::map(b1, |t| &t.0);
   |                        ^^^^^^^^

warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref.rs:11:13
   |
11 |     let b2: Ref<u32> = Ref::map(b1, |t| &t.0);
   |             ^^^^^^^^

warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref.rs:12:5
   |
12 |     assert_eq!(*b2, 5);
   |     ^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Linting on the `assert_eq!` seems pretty obviously wrong (what is the user supposed to do about that?), and I don't _think_ we should be linting `Ref::map`, either—it's just a static method [(no `self`)](https://doc.rust-lang.org/std/cell/struct.Ref.html#method.map) that we're accessing through `Ref`'s namespace, right?

_Also_, even when we do add the underscore-lifetime, the lints still go off (and point to the `'_`)—

```
zmd@ReflectiveCoherence:~/Code/Misc$ rustc +nightly elide_ref_fixed.rs 
warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref_fixed.rs:10:17
   |
10 |     let b1: Ref<'_, (u32, char)> = c.borrow();
   |                 ^^
   |
note: lint level defined here
  --> elide_ref_fixed.rs:2:9
   |
2  | #![warn(elided_lifetimes_in_paths)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref_fixed.rs:11:28
   |
11 |     let b2: Ref<'_, u32> = Ref::map(b1, |t| &t.0);
   |                            ^^^^^^^^

warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref_fixed.rs:11:17
   |
11 |     let b2: Ref<'_, u32> = Ref::map(b1, |t| &t.0);
   |                 ^^

warning: hidden lifetime parameters are deprecated, try `Foo<'_>`
  --> elide_ref_fixed.rs:12:5
   |
12 |     assert_eq!(*b2, 5);
   |     ^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```
